crates/rattler_repodata_gateway/src/gateway/sharded_subdir/token.rs

```diff
@@ -0,0 +1,163 @@
+                        let sender = Arc::new(sender);
+                        *token = PendingOrFetched::Pending(Arc::downgrade(&sender));
+                        Either::Left(sender)
+                    }
+                }
+            }
+        };
+
+        let sender = match sender_or_receiver {
+            Either::Left(sender) => sender,
+            Either::Right(mut receiver) => {
+                return match receiver.recv().await {
+                    Ok(Some(token)) => Ok(token),
+                    _ => {
+                        // If this happens the sender was dropped.
+                        Err(GatewayError::IoError(
+                            "a coalesced request for a token failed".to_string(),
+                            std::io::ErrorKind::Other.into(),
+                        ))
+                    }
+                };
+            }
+        };
+
+        let token_url = self
+            .token_base_url
+            .join("token")
+            .expect("invalid token url");
+        tracing::debug!("fetching token from {}", &token_url);
+
+        // Fetch the token
+        let token = {
+            let _permit = self.concurrent_request_semaphore.acquire().await;
+            let reporter = reporter.map(|r| (r, r.on_download_start(&token_url)));
+            let response = self
+                .client
+                .get(token_url.clone())
+                .header(CACHE_CONTROL, HeaderValue::from_static("max-age=0"))
+                .send()
+                .await
+                .and_then(|r| r.error_for_status().map_err(Into::into))
+                .map_err(GatewayError::from)?;
+
+            let bytes = response
+                .bytes_with_progress(reporter)
+                .await
+                .map_err(FetchRepoDataError::from)
+                .map_err(GatewayError::from)?;
+
+            if let Some((reporter, index)) = reporter {
+                reporter.on_download_complete(&token_url, index);
+            }
+
+            let token: Token = serde_json::from_slice(&bytes).map_err(|e| {
+                GatewayError::IoError("failed to parse sharded index token".to_string(), e.into())
+            })?;
+
+            Arc::new(token)
+        };
+
+        // Reacquire the token
+        let mut token_lock = self.token.lock();
+        *token_lock = PendingOrFetched::Fetched(Some(token.clone()));
+
+        // Publish the change
+        let _ = sender.send(Some(token.clone()));
+
+        Ok(token)
+    }
+}
+
+/// The token endpoint response.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Token {
+    pub token: Option<String>,
+    issued_at: Option<DateTime<Utc>>,
+    expires_in: Option<u64>,
+    pub shard_base_url: Option<Url>,
+}
+
+impl Token {
+    /// Returns true if the token is still considered to be valid.
+    pub fn is_fresh(&self) -> bool {
+        if let (Some(issued_at), Some(expires_in)) = (&self.issued_at, self.expires_in) {
```

Comment:
did we add thsi to the CEP?

Should we initialize `issued_at` with the current time if the token doesn't contain it?

Reply:
We did add it to the CEP yes: https://github.com/baszalmstra/ceps/blob/sharded-repodata/cep-16.md#authentication

Ill change the issued_at to the current time!